### PR TITLE
Initial support for OSCORE security object

### PR DESCRIPTION
**Pull request by rikard-sics**
Beginning to add support for the OSCORE security object.

Signed-off-by: Rikard Höglund <rikard.hoglund@ri.se>

### CODE REVIEW

File: leshan-client-core/src/main/java/org/eclipse/leshan/client/object/OSCORESecurity.java [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+ */
+public class OSCORESecurity extends BaseInstanceEnabler {
+
+    private static final Logger LOG = LoggerFactory.getLogger(Security.class);
+
+    byte[] OSCORE_master_secret;
+    String OSCORE_sender_id;
+    String OSCORE_recipient_id;
+    int OSCORE_AEAD_algorithm;
+    int OSCORE_HMAC_algorithm;
```
Comment (sbernard31) [inline leshan-client-core/src/main/java/org/eclipse/leshan/client/object/OSCORESecurity.java]: 1. Generally in Java, camelCase is prefered, so we do that in Leshan too.
One exception in Leshan, we are using snake case for test name.
So it should looks like `oscoreMasterSecret`.

2. Any reason do not use private access modifier ?

3. I know in the LWM2M specification,  resource name are prefixed by OSCORE and generally I prefer to keep as closest as possible as specification naming. But in this case I think it should be less verbose to just remove the `OSCORE` part in attribute name. So it looks like `masterSecret`, does it sound good to you ?

File: leshan-client-core/src/main/java/org/eclipse/leshan/client/object/OSCORESecurity.java [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+import org.eclipse.leshan.core.node.LwM2mResource;
+import org.eclipse.leshan.core.response.ReadResponse;
+import org.eclipse.leshan.core.response.WriteResponse;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * A simple {@link LwM2mInstanceEnabler} for the OSCORE Security (0) object.
+ */
+public class OSCORESecurity extends BaseInstanceEnabler {
```
Comment (sbernard31) [inline leshan-client-core/src/main/java/org/eclipse/leshan/client/object/OSCORESecurity.java]: The LWM2M object is named `LWM2M OSCORE`, should we name this java class `Oscore` ?

File: leshan-client-core/src/main/java/org/eclipse/leshan/client/object/OSCORESecurity.java [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+import org.eclipse.leshan.client.resource.BaseInstanceEnabler;
+import org.eclipse.leshan.client.resource.LwM2mInstanceEnabler;
+import org.eclipse.leshan.core.node.LwM2mResource;
+import org.eclipse.leshan.core.response.ReadResponse;
+import org.eclipse.leshan.core.response.WriteResponse;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * A simple {@link LwM2mInstanceEnabler} for the OSCORE Security (0) object.
```
Comment (sbernard31) [inline leshan-client-core/src/main/java/org/eclipse/leshan/client/object/OSCORESecurity.java]: LWM2M OSCORE object ID is 21, right ?

File: leshan-client-core/src/main/java/org/eclipse/leshan/client/object/OSCORESecurity.java [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+    public WriteResponse write(ServerIdentity identity, int resourceId, LwM2mResource value) {
+    	LOG.debug("Write on resource {}: {}", resourceId, value);
+    	// extend
+		return null;
+    }
+    
+    @Override
+    public ReadResponse read(ServerIdentity identity, int resourceid) {
+    	// extend
+		return null;
```
Comment (sbernard31) [inline leshan-client-core/src/main/java/org/eclipse/leshan/client/object/OSCORESecurity.java]: I suppose next step will be to implement this feature so it's not a big deal but just to let you know  the idea is more to return `ReadResponse.notFound()` or `ReadResponse.internalServerError("not implemented")` instead or return `null`

**Comment by rikard-sics** [comment]
Thanks for the feedback on the code. I will start going through it and adjust the pull request accordingly.

**Comment by rikard-sics** [comment]
Again, thank you for the feedback. I have now done 2 more commits where I believe I addressed all the points you listed.

File: leshan-client-core/src/main/java/org/eclipse/leshan/client/object/Oscore.java [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+    @Override
+    public WriteResponse write(ServerIdentity identity, int resourceId, LwM2mResource value) {
+        LOG.debug("Write on resource {}: {}", resourceId, value);
+        // extend
+        return WriteResponse.notFound();
+    }
+
+    @Override
+    public ReadResponse read(ServerIdentity identity, int resourceid) {
+        // extend
```
Comment (sbernard31) [inline leshan-client-core/src/main/java/org/eclipse/leshan/client/object/Oscore.java]: Any reason why you add log in write and not in read ?

File: leshan-client-core/src/main/java/org/eclipse/leshan/client/object/Oscore.java [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+        return ReadResponse.notFound();
+    }
+
+    public byte[] getOSCOREMasterSecret() {
```
Comment (sbernard31) [inline leshan-client-core/src/main/java/org/eclipse/leshan/client/object/Oscore.java]: As you renamed the attributes, you should rename getter, setter and setter parameters  too ?

Comment (rikard-sics) [reply leshan-client-core/src/main/java/org/eclipse/leshan/client/object/Oscore.java]: Yes, indeed I should do that.

**Comment by sbernard31** [comment]
It sounds good !  :)

Do you want, I integrate this is 2.0.x branch and you provide a new PR ?

**Comment by rikard-sics** [comment]
> It sounds good ! :)
> 
> Do you want, I integrate this is 2.0.x branch and you provide a new PR ?
> (I will squash all commits in one)

Yes, thanks. It would be great if you could do that. :)

**Comment by sbernard31** [comment]
Integrated in 2.0.x (all squashed in one commit bb2f1e5)

Thx @rikard-sics

**Comment by rikard-sics** [comment]
> Integrated in 2.0.x (all squashed in one commit [bb2f1e5](https://github.com/eclipse/leshan/commit/bb2f1e52cc8d35ad340cc1e828c58da861abd393))

Thanks a lot for all the feedback and assistance!

**Pull request closed without merging**